proto/iotics/api/input.proto

```diff
@@ -0,0 +1,203 @@
+  }
+
+  // DeleteInputResponse headers
+  Headers headers = 1;
+  // DeleteInputResponse payload
+  Payload payload = 2;
+}
+// ---------------------------------------
+
+// DescribeInputRequest is used to request the input metadata.
+message DescribeInputRequest {
+  // DescribeInputRequest arguments.
+  message Arguments {
+    // Input to describe
+    Input input = 1;
+    // HostID to describe a remote input (Optional, keep empty if input is local)
+    HostID remoteHostId = 2;
+  }
+
+  // DescribeInputRequest headers
+  Headers headers = 1;
+  // DescribeInputRequest mandatory arguments
+  Arguments args = 2;
+}
+
+// DescribeInputResponse provides metadata lookup for individual input resources.
+message DescribeInputResponse {
+  // DescribeInputResponse metadata result.
+  message MetaResult {
+    // Values semantically describing the Input messages
+    repeated Value values = 1;
+    // Custom properties associated with this input.
+    repeated Property properties = 2;
+  }
+  // DescribeInputResponse payload.
+  message Payload {
+    // Described input
+    Input input = 1;
+    // Metadata result
+    MetaResult result = 2;
+    // HostID of the described input. (Optional, empty if input is local)
+    HostID remoteHostId = 3;
+  }
+
+  // DescribeInputResponse headers
+  Headers headers = 1;
+  // DescribeInputResponse payload
+  Payload payload = 2;
+}
+// ---------------------------------------
+
+// UpsertInputWithMeta is used to describe the full input state. Used in UpsertTwinRequest.
+message UpsertInputWithMeta {
+  // Id of the input to create/update
+  string id = 1;
+  // Values to set
+  repeated Value values = 2;
+  // Properties to set
+  repeated Property properties = 3;
+}
+// ---------------------------------------
+
+// InputMessage describe a message that can be sent to an input
+message InputMessage {
+  // OccurredAt is the UTC timestamp of the message.
+  // Typically this is either the time at which an application sent this message
+  // or the time applicable to the message itself. (Optional: set to host time is not provided)
+  google.protobuf.Timestamp occurredAt = 1;
+  // Mime is the mime type of the encoded data.
+  string mime = 2;
+  // Data is the actual message, encoded according the the mime type. The data should follow the Input's
+  // value definitions but that is not enforced. (See also Value)
+  bytes data = 3;
+}
+// ---------------------------------------
+
+// ReceiveInputMessageRequest is used to receive messages sent to a given Input.
+message ReceiveInputMessageRequest {
+  // ReceiveInputMessageRequest arguments.
+  message Arguments {
+    // Input to listen messages from
+    Input input = 1;
+  }
+
+  // ReceiveInputMessageRequest headers
+  Headers headers = 1;
+  // ReceiveInputMessageRequest mandatory arguments
+  Arguments args = 2;
+}
+
+// ReceiveInputMessageResponse contains a message sent to the Input.
+message ReceiveInputMessageResponse {
+  message Payload {
```

Comment:
Should the originating host & sender twin be included here? e.g.:

```
// Twin which sent the input request
TwinID sender
// Host from which the input request came from
HostId hostId
```

Reply:
duplicate comment?

Reply:
> duplicate comment?

(Some GH client error.) Anyway, what do you think about this @smartrics ?

Reply:
it can certainly be useful for validation (as well as the clientref)  - to that extent we ought to think of a section with sender details including signature etc. in common with feeds

what other reasons do you think could be used for?

do you think they should go in payload or outside?

Reply:
@smartrics and I had a chat about this. We've decided it's worth waiting for initial input usage to see whether this is something that agents/clients would strongly benefit from. (Otherwise this can be tackled in conjunction with signing of requests at a later date.)